Editors/PathEditor.cpp

```diff
@@ -1,10 +1,238 @@
 #include "PathEditor.h"
+#include "Components/ArtManager.h"
+#include "Widgets/ResourceSelector.h"
 
 #include "ui_PathEditor.h"
 
+#include <QEvent>
+#include <QMouseEvent>
+#include <QResizeEvent>
+#include <QSpinBox>
+#include <QToolButton>
+
 PathEditor::PathEditor(ProtoModelPtr model, QWidget* parent) : BaseEditor(model, parent), ui(new Ui::PathEditor) {
   ui->setupUi(this);
-  connect(ui->actionSave, &QAction::triggered, this, &BaseEditor::OnSave);
+  connect(ui->saveButton, &QAbstractButton::pressed, this, &BaseEditor::OnSave);
+
+  ui->pathPreviewBackground->SetAssetView(ui->roomView);
+
+  // Prefer resizing the path view over the points editor
+  ui->splitter->setStretchFactor(0, 0);
+  ui->splitter->setStretchFactor(1, 1);
+
+  ui->mainToolBar->setStyleSheet("QToolBar{spacing:8px;}");
+
+  QLabel* xSnapLabel(new QLabel(tr("X Snap"), this));
+  ui->mainToolBar->addWidget(xSnapLabel);
+
+  QSpinBox* xSnap(new QSpinBox(this));
+  ui->mainToolBar->addWidget(xSnap);
+  xSnap->setRange(1, 99999999);
+
+  QLabel* ySnapLabel(new QLabel(tr("Y Snap"), this));
+  ui->mainToolBar->addWidget(ySnapLabel);
+
+  QSpinBox* ySnap(new QSpinBox(this));
+  ui->mainToolBar->addWidget(ySnap);
+  ySnap->setRange(1, 99999999);
+
+  ui->mainToolBar->addSeparator();
+
+  ResourceSelector* roomButton(new ResourceSelector(this, TreeNode::TypeCase::kRoom));
+  roomButton->setPopupMode(QToolButton::MenuButtonPopup);
+  roomButton->setIcon(ArtManager::GetIcon(":/resources/room.png"));
+  ui->mainToolBar->addWidget(roomButton);
+
+  roomLineEdit = new QLineEdit(this);
+  roomLineEdit->setReadOnly(true);
+  roomLineEdit->setMaximumWidth(150);
+  ui->mainToolBar->addWidget(roomLineEdit);
+
+  connect(roomButton, &QToolButton::pressed, [=]() {
+    roomLineEdit->setText("");
+    ui->roomView->SetResourceModel(nullptr);
+    pathModel->setData(Path::kBackgroundRoomNameFieldNumber, 0, "");
+  });
+
+  connect(roomButton->menu, &QMenu::triggered, this, [=](QAction* action) {
+    roomLineEdit->setText(action->text());
+    ui->roomView->SetResourceModel(MainWindow::resourceMap->GetResourceByName(TypeCase::kRoom, action->text())
+                                       ->GetSubModel(TreeNode::kRoomFieldNumber));
+    ui->pathPreviewBackground->SetZoom(1);
+    pathModel->setData(Path::kBackgroundRoomNameFieldNumber, 0, action->text());
+  });
+
+  ui->mainToolBar->addSeparator();
+
+  QToolButton* zoom(new QToolButton(this));
+  zoom->setIcon(ArtManager::GetIcon(":/actions/zoom.png"));
+  ui->mainToolBar->addWidget(zoom);
+  connect(zoom, &QToolButton::pressed, ui->pathPreviewBackground, [=]() { ui->pathPreviewBackground->SetZoom(1); });
+
+  QToolButton* zoomIn(new QToolButton(this));
+  zoomIn->setIcon(ArtManager::GetIcon(":/actions/zoom-in.png"));
+  ui->mainToolBar->addWidget(zoomIn);
+  connect(zoomIn, &QToolButton::pressed, ui->pathPreviewBackground,

```

Comment:
Shouldn't the background object just expose a function to handle zoom, instead of each editor hard-coding a default zoom factor?

Reply:
I didn't think it mattered for a 1 loc feature but I can probably inherit it I guess